src/Oracle.sol

```diff
@@ -0,0 +1,57 @@
+pragma solidity ^0.8.4;
+//SPDX-License-Identifier: MIT
+
+abstract contract Oracle {
+    error CardinalityNotAllowed();
+
+    struct Observation {
+        uint256 timestamp;
+        uint256 cumulativeWorth;
+    }
+
+    struct ObservationIndex {
+        uint128 cardinality;
+        uint128 lastIndex;
+    }
+
+    mapping(uint256 => Observation[65536]) observations;
+    mapping(uint256 => ObservationIndex) observationIndex;
+
+
+    function write(uint256 cloneId, uint256 price) internal {
```

Comment:
Should write be based on `protoId` instead of `cloneId`?

Reply:
why do you prefer tracking only `protoId`'s price?

Reply:
I'm just thinking that `protoId` is the actual identifier for the ERC20 and NFT pair. Specifying a `cloneId` would require the reader of the price oracle to know, or pass the index for the clone, when what they are likely interested in is how much the NFT itself is worth, not a specific clone's worth.

It would be nice to keep the oracle's price history when the index head moves to the next clone as well, which i think would be a challenge if each clone has its own independent oracle feed.